cmd/kubeadm/app/images/interface.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package images
+
+import (
+	"fmt"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// Puller is an interface for pulling images
+type Puller interface {
+	Pull(string) error
+}
+
+// Existence is an interface to determine if an image exists on the system
+// A nil error means the image was found
+type Existence interface {
+	Exists(string) error
+}
+
+// Images defines the set of behaviors needed for images realting to the CRI
+type Images interface {
+	Puller
+	Existence
+}
+
+// CRInterfacer is a struct that interfaces with the container runtime
+type CRInterfacer struct {
+	criSocket  string
+	exec       utilsexec.Interface
+	crictlPath string
+	dockerPath string
+}
+
+// NewCRInterfacer sets up and returns a CRInterfacer
+func NewCRInterfacer(execer utilsexec.Interface, criSocket string) (*CRInterfacer, error) {
+	var crictlPath, dockerPath string
+	var err error
+	if criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		if crictlPath, err = execer.LookPath("crictl"); err != nil {
+			return nil, fmt.Errorf("crictl is required for non docker container runtimes: %v", err)
```

Comment:
Are these errors a new addition?
They seem to halt the method preventing the new struct from being created.
It looks like you would not be able to use this with any other runtime than docker (and only with crictl also on the path).

Is this the expected behavior?

Reply:
The docker bit is a necessary hack that we'd like to eventually totally replace with crictl. We don't want to get into the business of writing this kind of logic for every container runtime tool, so we are only supporting crictl.

Also I have a ticket (holy cow i have a lot of tickets i need to do) to include crictl with the system packages so we can start depending on it more. 

The reason for the docker hack is that since the kubelet is the thing that hosts the cri dockershim and the kubelet isn't necessarily running by the time preflights run (I believe that is still true), we have to use the docker binary IF and only if we know we're using docker as a runtime (which we assert criSocket == default means we're using docker as the runtime). 

The longterm goal is to only use crictl.